scripts/build-reports.sh

```diff
@@ -0,0 +1,101 @@
+#!/usr/bin/env bash
+set -euo pipefail
+
+# Builds two benchmark reports for all benchmarked contracts:
+# - opcodes-report.json: ACIR opcode counts per function
+# - gates-report.json: circuit size (gates) per function
+#
+# For each subdirectory under contract-benchmarks:
+# 1) Run `nargo info` to generate target.
+# 2) Finds the single compiled Noir artifact JSON under "target/".
+# 3) Uses noir-inspector to list functions and filters those constrained functionswith opcodes > 1.
+# 4) Extracts each function into its own Noir artifact.
+# 5) Runs bb gates on the per-function artifact and collects metrics.
+
+ROOT_DIR=$(cd "$(dirname "$0")/.." && pwd)  # repo root
+CONTRACTS_DIR="$ROOT_DIR/contract-benchmarks"  # benchmarks root
+
+if [[ ! -d "$CONTRACTS_DIR" ]]; then
+  echo "Directory not found: $CONTRACTS_DIR" >&2
+  exit 1
+fi
+
+# Require noir-inspector once up-front
+if ! command -v noir-inspector >/dev/null 2>&1; then
+  echo "Error: noir-inspector not found in PATH" >&2
+  exit 1
+fi
+
+# Resolve bb backend once up-front (env BACKEND can override)
+BACKEND_BIN=${BACKEND:-"$HOME/.bb/bb"}
+if [[ ! -x "$BACKEND_BIN" ]]; then
+  echo "Error: bb not found or not executable at $BACKEND_BIN" >&2
+  exit 1
+fi
+
+METRICS_TMP=$(mktemp)  # temp file for metrics
+trap 'rm -f "$METRICS_TMP"' EXIT  
+
+for dir in "$CONTRACTS_DIR"/*/; do  # iterate over each contract
+  [[ -d "$dir" ]] || continue
+  CONTRACT_NAME=$(basename "$dir")
+
+  # Run nargo info inside the package directory if not found
+  if command -v nargo >/dev/null 2>&1; then
+    (cd "$dir" && nargo info)
+  else
+    echo "Warning: nargo not found in PATH; skipping 'nargo info' for $CONTRACT_NAME" >&2
+  fi
+
+  target_dir="$dir/target"
+  if [[ ! -d "$target_dir" ]]; then
+    echo "Target directory not found for $CONTRACT_NAME: $target_dir" >&2
+    continue
+  fi
```

Comment:
We can optimistically just read from the target directory if we assume that `nargo compile` succeeds.